pkg/operator/k8sutil/node.go

```diff
@@ -29,88 +29,49 @@ import (
 	"k8s.io/kubernetes/pkg/kubelet/apis"
 )
 
+// ValidNode returns true if the node (1) is schedulable, (2) meets Rook's placement terms, and
+// (3) is ready. False otherwise.
 func ValidNode(node v1.Node, placement rookalpha.Placement) (bool, error) {
-	// a node cannot be disabled
-	if node.Spec.Unschedulable {
+	if !GetNodeSchedulable(node) {
 		return false, nil
 	}
 
-	// a node matches the NodeAffinity configuration
-	// ignoring `PreferredDuringSchedulingIgnoredDuringExecution` terms: they
-	// should not be used to judge a node unusable
-	if placement.NodeAffinity != nil && placement.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution != nil {
-		nodeMatches := false
-		for _, req := range placement.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms {
-			nodeSelector, err := helper.NodeSelectorRequirementsAsSelector(req.MatchExpressions)
-			if err != nil {
-				return false, fmt.Errorf("failed to parse MatchExpressions: %+v, regarding as not match.", req.MatchExpressions)
-			}
-			if nodeSelector.Matches(labels.Set(node.Labels)) {
-				nodeMatches = true
-				break
-			}
-		}
-		if !nodeMatches {
-			return false, nil
-		}
+	p, err := NodeMeetsPlacementTerms(node, placement, false)
+	if err != nil {
+		return false, fmt.Errorf("failed to check if node meets Rook placement terms. %+v", err)
 	}
-
-	// a node is tainted and cannot be tolerated
-	for _, taint := range node.Spec.Taints {
-		isTolerated := false
-		for _, toleration := range placement.Tolerations {
-			if toleration.ToleratesTaint(&taint) {
-				isTolerated = true
-				break
-			}
-		}
-		if !isTolerated {
-			return false, nil
-		}
+	if !p {
+		return false, nil
 	}
 
-	// a node must be Ready
-	for _, c := range node.Status.Conditions {
-		if c.Type == v1.NodeReady {
-			return true, nil
-		}
+	if !NodeIsReady(node) {
+		return false, nil
 	}
-	logger.Infof("node %s is not ready. %+v", node.Name, node.Status.Conditions)
-	return false, nil
+
+	return true, nil
 }
 
+// GetValidNodes returns all nodes that (1) are not cordoned, (2) meet Rook's placement terms, and
```

Comment:
These are all the nodes used for adding nodes, right? Which would be a different set from nodes to consider for removal?

Reply:
Right. This code still needs some later refactoring IMO, but this keeps the existing definition of "valid" the same. It then rejects "valid" as a suitable state for determining if a node should be removed.